File: tools/field/field_glb.py
#!/usr/bin/env python3
"""Build the field's visual model, field.glb, from FIRST's Onshape export.

This is the field as it looks, beside `field.json`, which is the field as the simulator collides
it. They are built from different CAD by different means -- the collision model from FIRST's
published STEP, reduced to a convex shape per part; this from the Onshape assembly's tessellation,
which carries the real shape of every fillet and curve, and which STEP cannot carry the artwork
for. Keeping them apart is deliberate: a renderer wants the detail, and a physics engine wants
convex shapes, and one model cannot be good at both.

What they must agree about is where the field is, and `--check` holds them to it: a part both
name, in a different place in each, is a renderer drawing a flower where nothing collides.

    python3 tools/field/field_glb.py                    # fetch and build
    python3 tools/field/field_glb.py --export f.gltf    # build from an export already in hand

The export is a million and a half triangles, most of them spent on fillets that are invisible at
the size of a field. Points are snapped to a grid a twentieth of an inch across and the triangles
that collapse are dropped, which takes roughly three quarters of them and moves nothing further
than a twenty-third of an inch.

No dependencies beyond Python 3.
"""
import argparse
import json
import math
import os
import re
import sys

import gltf
import onshape
from step_to_field import clean

INCH = 0.0254
GRID_IN = 0.05
# The largest a landmark may differ between the two models before `--check` calls it a
# disagreement. The two reduce shape differently -- a hull of extreme points against a true
# tessellation -- so a diagonal member's centre honestly lands an inch or two apart.
AGREE_IN = 3.0
# Fewer landmarks than this and the check has not judged the field, whatever it reports.
LEAST_COMPARED = 10
OUT = os.path.join(os.path.dirname(__file__), '..', '..', 'TeamCode', 'src', 'test', 'resources',
                   'org', 'firstinspires', 'ftc', 'teamcode', 'sim', 'field.glb')
COLLISION = os.path.join(os.path.dirname(OUT), 'field.json')

# Hardware, and the parts the page draws itself. The collision model drops the same names; this
# keeps its own copy because the two models are allowed to diverge about what is worth drawing.
SKIP = re.compile(r'screw|fhts|nut\b|washer|rivet|rivnut|bolt|spacer|bearing|cable tie|plug|\bpin\b|'
                  r'damper|hinge|panel link|strap|clip|under tile|peanut|'
                  r'fastener|quick release|soft tiles|perimeter|rail|side glass', re.I)


def to_field(p):
    """The export's frame, metres -> Road Runner's, inches: a quarter turn about z.

    Onshape exports z up with the assembly's origin at the middle of the field, so the floor is
    already at z = 0 and only the turn is left. Which way it turns is settled by the one thing the
    field is not symmetric about: blue's hive stands where the export puts x positive, and this
    turn is the one that puts it at negative y, where `SimFieldTest` holds that it belongs.
    """
    return (p[1] / INCH, -p[0] / INCH, p[2] / INCH)


def snap(triangles, grid):
    """Triangles with their points snapped to a grid `grid` inches across, dropping those that
    come out with two corners in the same place.

    Crude beside a proper decimation, and that is the point: it moves no point further than half
    the grid's diagonal, whatever the shape, so the worst it can do is bounded and small. What it
    takes out is the tessellation of curves, which is where the triangles are.
    """
    out = []
    for a, b, c in triangles:
        at = tuple(tuple(round(v / grid) * grid for v in p) for p in (a, b, c))
        if at[0] == at[1] or at[1] == at[2] or at[0] == at[2]:
            continue
        out.append(at)
    return out


def visual_parts(parts, grid=GRID_IN):
    """The parts worth drawing, in the field frame, at the detail worth keeping."""
    out = []
    for part in parts:
        if any(SKIP.search(name) for name in list(part.path) + [part.name]):
            continue
        triangles = snap([tuple(to_field(p) for p in t) for t in part.triangles], grid)
        if triangles:
            out.append(gltf.Part(part.name, list(part.path), triangles, part.colour))
    return out


def _centre(part):
    points = [v for t in part.triangles for v in t]
    return tuple(sum(p[i] for p in points) / len(points) for i in range(3))


def disagreements(parts, collision, within=AGREE_IN, least=LEAST_COMPARED):
    """Where the visual model and the collision model disagree about the field itself.

    Not part by part: the two are built from different exports, which name and group their parts
    differently, and reconciling those names is a brittle way to ask a simple question. What is
    asked instead is the two things that go wrong when a frame is read wrong, and that nothing
    else would catch.

    Handedness. Blue's hive stands at negative y and red's at positive, and a quarter turn the
    wrong way, or a mirror, swaps them -- while leaving a field that looks perfectly reasonable,
    because everything else about it is symmetric.

    Scale. The model reaches as far as the collision model says the walls stand, which catches a
    frame read in metres, or in millimetres, or turned about the wrong axis.
    """
    blue = [_centre(p) for p in parts if re.search(r'blue', p.name, re.I)]
    red = [_centre(p) for p in parts if re.search(r'red', p.name, re.I)]
    if len(blue) < least or len(red) < least:
        return ['could not judge which way round the field is: %d blue parts and %d red, '
                'and at least %d of each are needed' % (len(blue), len(red), least)]

    off = []
    if not all(c[1] < 0 for c in blue) or not all(c[1] > 0 for c in red):
        off.append('the field is the wrong way round: blue belongs at negative y and red at '
                   'positive, and %d of %d blue parts and %d of %d red are on the wrong side'
                   % (sum(1 for c in blue if c[1] >= 0), len(blue),
                      sum(1 for c in red if c[1] <= 0), len(red)))

    # The field reaches past its own walls -- the trays stand off the ends, and the rows of game
    # pieces are set out beyond them -- so this is loose on the far side and tight on the near.
    # What it is for is a frame read in metres or millimetres, which is out by a factor of tens.
    half = collision['size'] / 2
    reach = max(abs(v[i]) for p in parts for t in p.triangles for v in t for i in (0, 1))
    if not half * 0.9 < reach < half * 3:
        off.append('the model reaches %.1f in from the middle; the collision model puts the walls '
                   'at %.1f in, so the frame or the units are wrong' % (reach, half))
    return off


def fetch():
    url = ('/api/v10/assemblies/d/' + onshape.FIELD_DOCUMENT + '/w/' + onshape.FIELD_WORKSPACE
           + '/e/' + onshape.FIELD_ASSEMBLY + '/gltf')
    return onshape.Client.configured().get(url, accept='model/gltf+json')


def main(argv):
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument('--export', help='a glTF export to build from, instead of fetching one')
    parser.add_argument('--grid', type=float, default=GRID_IN,
                        help='inches to snap points to (default: %(default)s)')
    parser.add_argument('--out', default=OUT)
    args = parser.parse_args(argv)

    if args.export:
        with open(args.export, 'rb') as f:
            export = f.read()
    else:
        try:
            export = fetch()
        except onshape.NoCredentials as refusal:
            print(refusal, file=sys.stderr)
            return 2
    try:
        read = gltf.read(export)
    except gltf.NotGltf as wrong:
        print('the export is not glTF this can read: %s' % wrong, file=sys.stderr)
        return 2

    parts = visual_parts(read, args.grid)
    if not parts:
        print('nothing was kept; the export named none of the parts we draw', file=sys.stderr)
        return 2

    with open(COLLISION) as f:
        collision = json.load(f)
    off = disagreements(parts, collision)
    if off:
        print('the visual model and the collision model disagree about where the field is:',
              file=sys.stderr)
        for line in off:
            print('  ' + line, file=sys.stderr)
        return 1

    body = gltf.write(parts)
    with open(args.out, 'wb') as f:
        f.write(body)
    before = sum(len(p.triangles) for p in read)
    after = sum(len(p.triangles) for p in parts)
    print('%s: %d parts, %d triangles of %d, %.1f MB'
          % (os.path.relpath(args.out), len(parts), after, before, len(body) / 1e6))
    return 0


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))

File: TeamCode/src/test/java/org/firstinspires/ftc/teamcode/sim/FieldGlb.java
package org.firstinspires.ftc.teamcode.sim;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;

public final class FieldGlb {
    public static final double GRID_IN = 0.05;
    public static final double NO_GRID = 0;
    public static final int LEAST_COMPARED = 10;

    /**
     * Which of the export's parts a model is made of. The field as it plays is not every part of the
     * assembly: two thirds of what the CAD holds is the hardware that holds the field together, which
     * is drawn nowhere and seen by nobody once the field is up. Full detail is the whole export, so
     * that "full" means what it says.
     */
    public enum Keep {
        WHAT_WE_DRAW,
        EVERY_PART
    }

    /**
     * How much of the CAD's own surface a model carries. Normals roughly double it -- there is one to
     * a vertex, the same size as the point it belongs to -- so the cheap model goes without them and
     * with a colour apiece, as it always has, and the expensive one carries what the CAD drew.
     */
    public enum Shading {
        A_COLOUR_APIECE,
        AS_THE_CAD_DREW_IT
    }

    private static final Pattern SKIP = Pattern.compile(
            "screw|fhts|nut\\b|washer|rivet|rivnut|bolt|spacer|bearing|cable tie|plug|\\bpin\\b|"
                    + "damper|hinge|panel link|strap|clip|under tile|peanut|"
                    + "fastener|quick release|soft tiles|perimeter|rail|side glass",
            Pattern.CASE_INSENSITIVE);

    private static final Pattern KEEP = Pattern.compile("field panel|ftc rail|side glass", Pattern.CASE_INSENSITIVE);

    private static final Pattern BLUE = Pattern.compile("blue", Pattern.CASE_INSENSITIVE);
    private static final Pattern RED = Pattern.compile("red", Pattern.CASE_INSENSITIVE);

    /** Points and the way they face, kept together because snapping may drop a triangle from both. */
    public record Mesh(double[] triangles, double[] normals) {}

    private FieldGlb() {}

    public static byte[] build(byte[] export, double fieldSizeIn, double grid, Keep keep, Shading shading) {
        List<Gltf.Part> parts = visualParts(Gltf.read(export), grid, keep, shading);
        if (parts.isEmpty()) {
            throw new IllegalStateException("nothing was kept; the export named none of the parts we draw");
        }
        List<String> off = disagreements(parts, fieldSizeIn);
        if (!off.isEmpty()) {
            throw new IllegalStateException(
                    "the visual model and the collision model disagree about where the field is:\n  "
                            + String.join("\n  ", off));
        }
        return Gltf.write(parts);
    }

    public static List<Gltf.Part> visualParts(List<Gltf.Part> parts, double grid, Keep keep, Shading shading) {
        boolean asDrawn = shading == Shading.AS_THE_CAD_DREW_IT;
        List<Gltf.Part> out = new ArrayList<>();
        for (Gltf.Part part : parts) {
            if (keep == Keep.WHAT_WE_DRAW && !KEEP.matcher(part.name).find() && droppedByName(part)) {
                continue;
            }
            double[] facing = asDrawn && part.normals != null ? toFacing(part.normals) : null;
            Mesh kept = snap(toField(part.triangles), facing, grid);
            if (kept.triangles().length > 0) {
                out.add(new Gltf.Part(
                        part.name,
                        part.path,
                        kept.triangles(),
                        kept.normals(),
                        asDrawn || part.colour == null ? part.material : Gltf.colouredMaterial(part.colour)));
            }
        }
        return out;
    }

    private static boolean droppedByName(Gltf.Part part) {
        if (SKIP.matcher(part.name).find()) {
            return true;
        }
        for (String above : part.path) {
            if (SKIP.matcher(above).find()) {
                return true;
            }
        }
        return false;
    }

    /**
     * The frame change as a normal feels it: the same quarter turn, and none of the inch. A normal
     * says which way a face points, so turning it is right and scaling it is not.
     */
    public static double[] toFacing(double[] normals) {
        double[] out = new double[normals.length];
        for (int i = 0; i < normals.length; i += 3) {
            out[i] = normals[i + 1];
            out[i + 1] = -normals[i];
            out[i + 2] = normals[i + 2];
        }
        return out;
    }

    public static double[] toField(double[] points) {
        double[] out = new double[points.length];
        for (int i = 0; i < points.length; i += 3) {
            out[i] = points[i + 1] / Onshape.INCH_IN_METRES;
            out[i + 1] = -points[i] / Onshape.INCH_IN_METRES;
            out[i + 2] = points[i + 2] / Onshape.INCH_IN_METRES;
        }
        return out;
    }

    public static Mesh snap(double[] triangles, double[] normals, double grid) {
        double[] kept = new double[triangles.length];
        double[] keptFacing = normals == null ? null : new double[normals.length];
        int at = 0;
        for (int i = 0; i < triangles.length; i += 9) {
            double[] corner = new double[9];
            for (int c = 0; c < 9; c++) {
                corner[c] = grid > 0 ? Math.rint(triangles[i + c] / grid) * grid : triangles[i + c];
            }
            if (same(corner, 0, 3) || same(corner, 3, 6) || same(corner, 0, 6)) {
                continue;
            }
            System.arraycopy(corner, 0, kept, at, 9);
            if (keptFacing != null) {
                System.arraycopy(normals, i, keptFacing, at, 9);
            }
            at += 9;
        }
        double[] out = new double[at];
        System.arraycopy(kept, 0, out, 0, at);
        double[] facing = null;
        if (keptFacing != null) {
            facing = new double[at];
            System.arraycopy(keptFacing, 0, facing, 0, at);
        }
        return new Mesh(out, facing);
    }

    private static boolean same(double[] corner, int a, int b) {
        return corner[a] == corner[b] && corner[a + 1] == corner[b + 1] && corner[a + 2] == corner[b + 2];
    }

    public static List<String> disagreements(List<Gltf.Part> parts, double fieldSizeIn) {
        List<double[]> blue = new ArrayList<>();
        List<double[]> red = new ArrayList<>();
        for (Gltf.Part part : parts) {
            if (BLUE.matcher(part.name).find()) {
                blue.add(middleOf(part));
            }
            if (RED.matcher(part.name).find()) {
                red.add(middleOf(part));
            }
        }

        List<String> off = new ArrayList<>();
        if (blue.size() < LEAST_COMPARED || red.size() < LEAST_COMPARED) {
            off.add("could not judge which way round the field is: " + blue.size() + " blue parts and " + red.size()
                    + " red, which is fewer than the " + LEAST_COMPARED + " it takes to tell");
            return off;
        }

        long blueWrong = blue.stream().filter(centre -> centre[1] >= 0).count();
        long redWrong = red.stream().filter(centre -> centre[1] <= 0).count();
        if (blueWrong > 0 || redWrong > 0) {
            off.add(String.format(
                    Locale.ROOT,
                    "the field is the wrong way round: blue belongs at negative y and red at positive, and %d of %d "
                            + "blue parts and %d of %d red are on the wrong side",
                    blueWrong,
                    blue.size(),
                    redWrong,
                    red.size()));
        }

        double half = fieldSizeIn / 2;
        double reach = 0;
        for (Gltf.Part part : parts) {
            for (int i = 0; i < part.triangles.length; i += 3) {
                reach = Math.max(reach, Math.abs(part.triangles[i]));
                reach = Math.max(reach, Math.abs(part.triangles[i + 1]));
            }
        }
        if (!(half * 0.9 < reach && reach < half * 3)) {
            off.add(String.format(
                    Locale.ROOT,
                    "the model reaches %.1f in from the middle; the collision model puts the walls at %.1f in, so "
                            + "the frame or the units are wrong",
                    reach,
                    half));
        }
        return off;
    }

    private static double[] middleOf(Gltf.Part part) {
        double[] sum = new double[3];
        int points = part.triangles.length / 3;
        for (int i = 0; i < part.triangles.length; i += 3) {
            sum[0] += part.triangles[i];
            sum[1] += part.triangles[i + 1];
            sum[2] += part.triangles[i + 2];
        }
        return new double[] {sum[0] / points, sum[1] / points, sum[2] / points};
    }
}
